From be2760b1358dca3199de2c31755c80102f70761c Mon Sep 17 00:00:00 2001
From: Anson Huang <b20788@freescale.com>
Date: Mon, 27 Jun 2011 11:35:39 +0800
Subject: [PATCH 588/670] ENGR00152131 [MX28]Suspend/resume defect

1.We should gate all ref_clk that sourced from
PLL before power down PLL, or the ref clock
will fail when PLL is power up.

2.Move the temp stack to front of the function,
or when we add/del some code, the addr of this
temp stack may cause build fail due to ARM const
compile rule.

Signed-off-by: Anson Huang <b20788@freescale.com>
---
 arch/arm/mach-mx28/sleep.S | 53 ++++++++++++++++++++++++++++++++++++++--------
 1 file changed, 44 insertions(+), 9 deletions(-)

diff --git a/arch/arm/mach-mx28/sleep.S b/arch/arm/mach-mx28/sleep.S
index 13a376c..61df79c 100644
--- a/arch/arm/mach-mx28/sleep.S
+++ b/arch/arm/mach-mx28/sleep.S
@@ -35,6 +35,8 @@
 	(MX28_SOC_IO_ADDRESS(CLKCTRL_PHYS_ADDR) + HW_CLKCTRL_CPU)
 #define HW_CLKCTRL_HBUS_ADDR \
 	(MX28_SOC_IO_ADDRESS(CLKCTRL_PHYS_ADDR) + HW_CLKCTRL_HBUS)
+#define HW_CLKCTRL_FRAC0_ADDR \
+	(MX28_SOC_IO_ADDRESS(CLKCTRL_PHYS_ADDR) + HW_CLKCTRL_FRAC0)
 #define HW_POWER_MINPWR_ADDR \
 	(MX28_SOC_IO_ADDRESS(POWER_PHYS_ADDR) + HW_POWER_MINPWR)
 #define HW_POWER_RESET_ADDR \
@@ -152,6 +154,10 @@ ENTRY(mx28_cpu_standby)
 	@ save registers on stack
 	stmfd	sp!, {r0 - r9, lr}
 
+	.space	0x100
+__mx28_temp_stack:
+	.space	128
+
 	adr	r9, __mx28_temp_stack
 
 	@ clean cache
@@ -174,7 +180,7 @@ ENTRY(mx28_cpu_standby)
 	cmp	r0, #0
 	bne	11b
 
-	@ gate clk
+	@ gate EMI clk
 	mov	r0, #(HW_CLKCTRL_EMI_ADDR & 0x000000FF)
 	orr	r0, r0, #(HW_CLKCTRL_EMI_ADDR & 0x0000FF00)
 	orr	r0, r0, #(HW_CLKCTRL_EMI_ADDR & 0x00FF0000)
@@ -183,9 +189,23 @@ ENTRY(mx28_cpu_standby)
 	orr	r1, r1, #(BM_CLKCTRL_EMI_CLKGATE)
 	str	r1, [r0]
 
-//	PM_SET_AND_BACKUP_REG HW_PINCTRL_EMI_DS_CTRL_ADDR,\
-//	BM_PINCTRL_EMI_DS_CTRL_DDR_MODE,\
-//	BF_PINCTRL_EMI_DS_CTRL_DDR_MODE(0x1), 4
+	@ gate ref_emi clk
+	mov	r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x000000FF)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x0000FF00)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x00FF0000)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0xFF000000)
+	ldr	r1, [r0]
+	orr	r1, r1, #(BM_CLKCTRL_FRAC0_CLKGATEEMI)
+	str	r1, [r0]
+
+	@ gate ref_cpu clk
+	mov	r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x000000FF)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x0000FF00)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x00FF0000)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0xFF000000)
+	ldr	r1, [r0]
+	orr	r1, r1, #(BM_CLKCTRL_FRAC0_CLKGATECPU)
+	str	r1, [r0]
 
 	mov	r0, #(HW_PINCTRL_EMI_DS_CTRL_ADDR & 0x000000FF)
 	orr	r0, r0, #(HW_PINCTRL_EMI_DS_CTRL_ADDR & 0x0000FF00)
@@ -371,7 +391,25 @@ ENTRY(mx28_cpu_standby)
 	orr	r1, r1, #(BF_PINCTRL_EMI_DS_CTRL_DDR_MODE(3))
 	str	r1, [r0]
 
-	@ ungate clk
+	@ ungate ref_cpu clk
+	mov	r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x000000FF)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x0000FF00)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x00FF0000)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0xFF000000)
+	ldr	r1, [r0]
+	bic	r1, r1, #(BM_CLKCTRL_FRAC0_CLKGATECPU)
+	str	r1, [r0]
+
+	@ ungate ref_emi clk
+	mov	r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x000000FF)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x0000FF00)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0x00FF0000)
+	orr	r0, r0, #(HW_CLKCTRL_FRAC0_ADDR & 0xFF000000)
+	ldr	r1, [r0]
+	bic	r1, r1, #(BM_CLKCTRL_FRAC0_CLKGATEEMI)
+	str	r1, [r0]
+
+	@ ungate EMI clk
 	mov	r0, #(HW_CLKCTRL_EMI_ADDR & 0x000000FF)
 	orr	r0, r0, #(HW_CLKCTRL_EMI_ADDR & 0x0000FF00)
 	orr	r0, r0, #(HW_CLKCTRL_EMI_ADDR & 0x00FF0000)
@@ -380,7 +418,6 @@ ENTRY(mx28_cpu_standby)
 	bic	r1, r1, #(BM_CLKCTRL_EMI_CLKGATE)
 	str	r1, [r0]
 
-//	PM_SET_RESTORE_REG HW_PINCTRL_EMI_DS_CTRL_ADDR, 4
 	@ restore normal DRAM mode
 	mov	r0, #(HW_DRAM_CTL22_ADDR & 0x000000FF)
 	orr	r0, r0, #(HW_DRAM_CTL22_ADDR & 0x0000FF00)
@@ -389,6 +426,7 @@ ENTRY(mx28_cpu_standby)
 	ldr	r1, [r0]
 	bic	r1, r1, #(BF_DRAM_CTL22_LOWPOWER_CONTROL(2))
 	str	r1, [r0]
+
 	@ wait for it to actually happen
 	mov	r0, #24 << 12
 12:	sub	r0, r0, #1
@@ -402,9 +440,6 @@ ENTRY(mx28_cpu_standby)
 	@ restore regs and return
 	ldmfd   sp!, {r0 - r9, pc}
 
-	.space	0x100
-__mx28_temp_stack:
-	.space	128
 
 #ifdef CONFIG_STMP378X_RAM_FREQ_SCALING
 #include "emi.inc"
-- 
1.8.0

